New decomposition of `QubitUnitary` based on Cartan decomposition

dwierichs opened the pull request:
**Context:**
See bug report. This decomposition has been a bit fragile through time.

**Description of the Change:**
New implementation for 3CNOT case.

**Benefits:**
Correct code.

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**
Fixes #7467 and #7534

[sc-91723]
[sc-92318]

dwierichs commented:
@Alex-Preciado The PR is not review-ready yet because it misses docs.

astralcai commented on pennylane/ops/op_math/decompositions/unitary_decompositions.py:
```diff
@@ -616,94 +618,124 @@ def _decompose_2_cnots(U, wires, initial_phase):
-    gammaU = math.dot(u, math.T(u))
-    # !Note: [sc-89460]
-    evs, _ = math.linalg.eig(gammaU)
+def ai_kak(u):
```
Can you make the helper functions private?

astralcai commented:
Can you also update `_two_qubit_resource` to reflect the resource requirement of the new algorithm? Also since `_decompose_2_cnots` does not guarantee correctness either, you can probably remove that as well, and update `two_qubit_decomp_rule` to dispatch to `_decompose_3_cnots` for both the 2 CNOT case and 3 CNOT case.

dwierichs commented on pennylane/ops/op_math/decompositions/unitary_decompositions.py:
```diff
@@ -616,94 +618,124 @@ def _decompose_2_cnots(U, wires, initial_phase):
-     -╭U- = -C--╭X--RZ(d)--╭C---------╭X--A-
-     -╰U- = -D--╰C--RY(b)--╰X--RY(a)--╰C--B-
-    """
+def _multidot(*matrices):
```
`qml.math.linalg.multi_dot` dispatches to `np.linalg.multi_dot` but is not interface agnostic (doesn't work with torch, for example).

dwierichs replied:
For JAX, `math.linalg.multi_dot` also seems to dispatch to `numpy`, causing an error when tracing.

Regarding the performance: We're talking about 4x4 matrices and the rest of the file calls `math.dot` repeatedly as well. I'm happy to look into making this better, but maybe it's not part of the P0 issue :)

astralcai commented on pennylane/ops/op_math/decompositions/unitary_decompositions.py:
```diff
@@ -561,149 +480,210 @@ def _decompose_1_cnot(U, wires, initial_phase):
+    gphase = _decompose_0_cnots(M_2, wires, 0.0)
+    _central_circuit(a, b, d, wires)
+    gphase += _decompose_0_cnots(M_1, wires, gphase)
+    # gphase is zero because we are guaranteed that M_1 and M_2 have unit determinant
```
If we do expect the `gphase` from `M_2` and `M_1` to cancel out here, it may not be necessary to actually accumulate it.

astralcai commented on tests/ops/op_math/test_decompositions.py:
```diff
@@ -984,13 +984,14 @@ def test_two_qubit_decomposition_3_cnots(self, U, wires):
     @pytest.mark.parametrize("U", samples_2_cnots)
     def test_two_qubit_decomposition_2_cnots(self, U, wires):
         """Test that a two-qubit matrix using 2 CNOTs isolation is correctly decomposed."""
+        # NOTE: Currently, we defer to the 3-CNOTs function for the 2-CNOTs case.
 
         U = qml.math.convert_to_su4(np.array(U))
 
         assert _compute_num_cnots(U) == 2
 
         obtained_decomposition = two_qubit_decomposition(U, wires=wires)
-        assert len(obtained_decomposition) == 9
+        assert len(obtained_decomposition) == 11  #  9
 
         tape = qml.tape.QuantumScript(obtained_decomposition)
         obtained_matrix = qml.matrix(tape, wire_order=wires)
```
Honestly just combine the test matrices used here with the test case for the 3-cnot case, and remove this test function.

dwierichs replied:
Shouldn't we just keep these tests around so we can easily modify them back into two-CNOT mode once we have such a function? :)

astralcai replied:
Oh wait the original 2-CNOT case has been removed. Yeah nvm in that case you can leave this as is, although maybe you can slightly explain what this `# 9` comment means.

josh146 commented on doc/releases/changelog-dev.md:
```diff
@@ -402,6 +402,11 @@
 
 <h3>Breaking changes 💔</h3>
 
+* Decompositions of two-qubit unitaries via `two_qubit_decomposition` now use 3 CNOTs even if
+  just 2 CNOTs would suffice theoretically. This is the price for guaranteeing correctness in
+  all edge cases, and will be improved in a later contribution.
+  [(#7474)](https://github.com/PennyLaneAI/pennylane/pull/7474)
```
I think it's worth erring on the side of breaking change here, since for some users the explicit decomposition _is_ important, and if that changed they may consider it a breaking change.

So I would suggest keeping it here for visibility, but emphasizing:

- Yes decomposition will change, this may/may not affect you depending on your workflow
- However, this does fix a bug which resulted in incorrect results previously
- Overall, numerical results should stay the same.

dwierichs replied:
I would be a bit annoyed if my gate counts grew silently and I could not find out in the changelog why that might be. If a user's workflow did not hit any of the edge cases but was using the two-CNOT scenario, this PR makes things strictly worse for them, no?

JerryChen97 reviewed (dismissed):
Could we directly add the special matrix
```python
U = np.array(
    [[-0.40836553 - 0.28850926j, 0.23864466 - 0.4393731j,
      0.02190401 + 0.06284322j, -0.23169726 - 0.66474606j],
     [-0.48865844 - 0.10589111j, 0.49770294 + 0.04787262j,
      -0.02177958 + 0.45019526j, -0.02632749 + 0.54420285j],
     [0.49276489 - 0.08475118j, 0.47943096 - 0.14193644j,
      0.51448332 - 0.17932317j, -0.42560958 + 0.14834623j],
     [0.42041002 - 0.27065738j, -0.31807806 - 0.38578018j,
      -0.03401683 + 0.70314565j, -0.00321586 + 0.06647341j]]
)  # ref:  https://github.com/PennyLaneAI/pennylane/issues/7467
```
to wherever it makes sense in our test suite? Also keep the reference to the issue as well so that we still remember even years later.

dwierichs commented on pennylane/ops/op_math/decompositions/unitary_decompositions.py:
```diff
@@ -561,149 +480,210 @@ def _decompose_1_cnot(U, wires, initial_phase):
-    # Now we find the A, B, C, D in SU(2), and return the decomposition
-    A, B, C, D = _extract_su2su2_prefactors(U, V)
+def _extract_abde(A):
+    """Extract the parameters for the central part of a 3-CNOT circuit as well
```
There is no reference for this step as far as I am aware.

dwierichs commented on pennylane/ops/op_math/decompositions/unitary_decompositions.py:
```diff
@@ -561,149 +480,210 @@ def _decompose_1_cnot(U, wires, initial_phase):
-    where V is as depicted in the circuit below:
-     -╭U- = -C--╭X--RZ(d)--╭C---------╭X--A-
-     -╰U- = -D--╰C--RY(b)--╰X--RY(a)--╰C--B-
+    """Decompose a unitary 4x4 matrix into a 3-CNOT circuit.
```
Added a reference for the AI Cartan decomposition. The rest is in the docstring, I'd say.

dwierichs commented:
Very good suggestion, @JerryChen97!
I added the test case and uncovered another instability that is cause by the newest version. In order to circumvent the
problem of degenerate eigenvalues of `imag(Delta) + real(Delta)`, I replaced it by `0.1 real(Delta) + 10 imag(Delta)`.
This causes the eigenvalues to separate in scale and effectively allows us to simultaneously diagonalize `real(Delta)` and `imag(Delta)`.
Basically, if we do not do this, even though the real and imaginary parts share an eigenbasis, we might not be computing it due to eigenvalue degeneracies.

astralcai commented on pennylane/ops/op_math/decompositions/unitary_decompositions.py:
```diff
@@ -561,149 +480,257 @@ def _decompose_1_cnot(U, wires, initial_phase):
-    return math.cast_like(-np.pi / 4, initial_phase)
+    a, b, d, e = _extract_abde(A_M)
+
+    _ = _decompose_0_cnots(M_2, wires, 0.0)
+    _central_circuit(a, b, d, wires)
+    _ = _decompose_0_cnots(M_1, wires, 0.0)
```
```suggestion
    _decompose_0_cnots(M_2, wires, 0.0)
    _central_circuit(a, b, d, wires)
    _decompose_0_cnots(M_1, wires, 0.0)
```

isaacdevlugt commented on doc/releases/changelog-dev.md:
```diff
@@ -413,6 +413,12 @@
 
 <h3>Breaking changes 💔</h3>
 
+* Decompositions of two-qubit unitaries via `two_qubit_decomposition` now use 3 CNOTs even if
```
Is `two_qubit_decomposition` a user-facing function? 🤔

astralcai replied:
It is public and visible in the documentation, not sure if anyone actually uses it though.

dwierichs replied:
The [compilation hub entry on two-qubit synthesis](https://pennylane.ai/compilation/two-qubit-synthesis) advertises it, for example.

isaacdevlugt commented on doc/releases/changelog-dev.md:
```diff
@@ -413,6 +413,12 @@
 <h3>Breaking changes 💔</h3>
 
+* Decompositions of two-qubit unitaries via `two_qubit_decomposition` now use 3 CNOTs even if
+  just 2 CNOTs would suffice theoretically. This may or may not affect a user's workflow, depending
+  on the numerical matrices provided to the decomposition. This change fixes a previous bug that
+  affected the correctness of the decomposition in some edge cases.
```
I'm still a little confused here 😵. If 2 CNOTs suffice, then why was there a bug? Perhaps this is a better wording (less words 😅).

```suggestion
  just 2 CNOTs would suffice theoretically. This fixes a bug in the implementation for this decomposition
  that affected its correctness in some edge cases.
```

dwierichs replied:
scratch that, given your other feedback I rewrote it.

isaacdevlugt commented on doc/releases/changelog-dev.md:
```diff
@@ -553,6 +559,12 @@ Here's a list of deprecations made this release. For a more detailed breakdown o
 
 <h3>Bug fixes 🐛</h3>
 
+* Implemented a new decomposition of two-qubit `QubitUnitary` operators in `two_qubit_decomposition`
+  based on a type-AI Cartan decomposition. It fixes previously faulty edge cases for unitaries
+  that require 2 or 3 CNOT gates. Now, 3 CNOTs are used for both cases, using one more
+  CNOT than theoretically required in the former case.
+  [(#7474)](https://github.com/PennyLaneAI/pennylane/pull/7474)
```
```suggestion
* A new decomposition was Implemented for two-qubit `QubitUnitary` operators in `two_qubit_decomposition`
  based on a type-AI Cartan decomposition. It fixes previously faulty edge cases for unitaries
  that require 2 or 3 CNOT gates. Now, 3 CNOTs are used for both cases, using one more
  CNOT than theoretically required in the former case.
  [(#7474)](https://github.com/PennyLaneAI/pennylane/pull/7474)
```

Couple comments on this outside of the suggestion:
- Is there a link we can provide for type-AI Cartan decomposition?
- I found this bug-fix entry to be a lot clearer than the entry in breaking changes 🤔.

JerryChen97 commented on pennylane/ops/op_math/decompositions/unitary_decompositions.py:
```diff
@@ -561,149 +480,257 @@ def _decompose_1_cnot(U, wires, initial_phase):
-        lambda: math.angle(evs[2]),
-        lambda: math.angle(evs[1]),
+    Delta = math.dot(U, math.transpose(U))
+    # Delta^T=Delta implies that Delta, real(Delta), and imag(Delta) share an eigenbasis
```
Let's also denote that `U` is unitary here

JerryChen97 commented on pennylane/ops/op_math/decompositions/unitary_decompositions.py:
```diff
@@ -561,149 +480,257 @@ def _decompose_1_cnot(U, wires, initial_phase):
-        lambda: math.angle(evs[2]),
-        lambda: math.angle(evs[1]),
+    Delta = math.dot(U, math.transpose(U))
+    # Delta^T=Delta implies that Delta, real(Delta), and imag(Delta) share an eigenbasis
```
```suggestion
    # U is unitary, so Delta = U U^T is also unitary (product of unitaries).
    # Since Delta is unitary, it is normal (commutes with its adjoint: Delta Delta† = Delta† Delta).
    # For a normal matrix, real(Delta) and imag(Delta) commute with each other and with Delta,
    # therefore they share a common eigenbasis.
```

dwierichs replied:
I made this argument more specific to the unitary case, hope that's okay, too.

JerryChen97 commented on pennylane/ops/op_math/decompositions/unitary_decompositions.py:
```diff
@@ -561,149 +480,257 @@ def _decompose_1_cnot(U, wires, initial_phase):
 
-    x = math.angle(evs[0])
+def _ai_kak(U):
+    """Compute a type-AI Cartan decomposition of ``U`` in the standard basis/representation.
```
```suggestion
    """Compute a type-AI Cartan decomposition of a unitary ``U`` in the standard basis/representation.
```
